src/ol/expr/cpu.js

```diff
@@ -92,14 +85,7 @@ export function newEvaluationContext() {
  * @return {ExpressionEvaluator} The expression evaluator.
  */
 export function buildExpression(encoded, type, context) {
-  const expression = parse(encoded, context);
-  if (!overlapsType(type, expression.type)) {
-    const expected = typeName(type);
-    const actual = typeName(expression.type);
-    throw new Error(
-      `Expected expression to be of type ${expected}, got ${actual}`,
-    );
-  }
+  const expression = parse(encoded, type, context);
```

Comment:
Parsing used to try to infer types from literal values and then assert that the expected type overlaps with the inferred type.  The new strategy is to pass down the expected type from the top - casting or coercing values to the expected type where appropriate.